accepted/2021/package-validation/apicompat.md

```diff
@@ -0,0 +1,143 @@
+
+## How is this tool going to be consumed
+
+We are proposing a tool that can be consumed as a Roslyn analyzer and also providing a dotnet tool that people can integrate to their build systems as an out of proc tool or run directly from the console. 
+
+Something to consider would be adding an MSBuild Task entrypoint so that libraries that need to run it in their build, they don't need to shell out a process to invoke it and just run in-proc.
+
+## Tool input scenarios
+
+### 1. Text files
+
+Just as `Microsoft.CodeAnalysis.PublicAnalyzers` currently accepts a `PublicAPI.Unshipped.txt` and `Public.Shipped.txt` files as inputs, we should preserve this behavior, but extend some of the rules and rule settings that we support so that all inputs have equal support.
+
+### 2. Binary inputs
+
+Having binaries as inputs is really helpful for compatibility, as this allows you to run a reference assembly against multiple runtime implementations (i.e win-x64, linux-x64, etc) and make sure you are compatible across all runtimes with previous versions of your library and also with the reference assembly.
+
+### 3. Source inputs
+
+Instead of binaries or text files, it might be helpful to capture what the contract is as a C# minimal file (somewhat like a native header file) that contains the API definitions and metadata. For this we would provide a separate tool, aka `GenAPI` that given an binary input produces minimal C# code. This tool will also be Roslyn based, protype can be found here: https://github.com/safern/roslyn-genapi. This provides a contract that is "easier" to read and is also buildable and produces a minimal reference assembly, currently what `dotnet/runtime` uses as a contract.
+
+## Compatibility Rules
+
+After looking at the existing tools, these is an overview of the proposed rules for the first version of the tool and that will run by default.
+
+### 1. Attribute Difference
+This rules checks attributes on public members and gives an error if there is a miss match on the attributes. This rule checks attributes on:
+- Types (Class, Struct, Enum, Interface, Record, Delegate)
+- Fields
+- Properties
+- Methods
+- Event
+- ReturnValue
+- Constructor
+- Generic Parameters
+
+Should we check for assembly attributes maybe as opt-in?
+
+It should check for the following:
+1. Attribute does not exist in the contract or implementation.
+2. Attribute arguments does not match in both the contract and implementation.
+    2.1 For flag arguments (i.e AtrributeTargets) we should decide whether expanding or removing a flag is breaking or not.
+
+We should allow a Rule setting to disable attribute diffing for a list of attributes where libraries can add attributes they don't care about compat, i.e `EditorBrowsableAttribute`. 
```

Comment:
Would we be opinionated by default, e.g. with a built-in default list of attributes we believe shouldn't impact compatibility?  Or would we require every developer to build that list up themselves?

Reply:
Yeah, I think we should have a default list of attributes that shouldn't impact compatibility, like compiler generated attributes, obsolete attributes, editor browsable, etc. Will add a note about that here.